pkg/controller/appinst_zone_test.go

```diff
@@ -0,0 +1,216 @@
+	}
+	_, err = apis.clusterInstApi.store.Put(ctx, &reservD, sync.SyncWait)
+	require.Nil(t, err)
+	_, err = apis.clusterInstApi.store.Put(ctx, &reservK, sync.SyncWait)
+	require.Nil(t, err)
+	// cloudlet 0 should have default MT cluster
+	defaultMT0Key := cloudcommon.GetDefaultMTClustKey(cloudletData[0].Key)
+	var defaultMT0 edgeproto.ClusterInst
+	found := apis.clusterInstApi.store.Get(ctx, defaultMT0Key, &defaultMT0)
+	require.True(t, found)
+	// cloudlet 4 is a single k8s cluster platform
+	singleMT4Key := cloudcommon.GetDefaultClustKey(cloudletData[4].Key, cloudletData[4].SingleKubernetesClusterOwner)
+	var singleMT4 edgeproto.ClusterInst
+	found = apis.clusterInstApi.store.Get(ctx, singleMT4Key, &singleMT4)
+	require.True(t, found)
+
+	appKey := edgeproto.AppKey{
+		Organization: "devorg",
+		Name:         "testapp",
+		Version:      "1.0",
+	}
+	getApp := func() *edgeproto.App {
+		return &edgeproto.App{
+			Key:             appKey,
+			ImageType:       edgeproto.ImageType_IMAGE_TYPE_DOCKER,
+			AccessPorts:     "tcp:443",
+			AccessType:      edgeproto.AccessType_ACCESS_TYPE_LOAD_BALANCER,
+			DefaultFlavor:   flavorData[0].Key,
+			AllowServerless: false,
+			ServerlessConfig: &edgeproto.ServerlessConfig{
+				Vcpus: *edgeproto.NewUdec64(0, 500*edgeproto.DecMillis),
+				Ram:   20,
+			},
+		}
+	}
+	getAppInst := func() *edgeproto.AppInst {
+		return &edgeproto.AppInst{
+			Key: edgeproto.AppInstKey{
+				Name:         "testinst",
+				Organization: "devorg",
+			},
+			AppKey:  appKey,
+			ZoneKey: zone.Key,
+			Flavor:  flavorData[0].Key,
+		}
+	}
+	cctx := DefCallContext()
+
+	verifyCloudlets := func(pcs []*potentialInstCloudlet, idxs ...int) {
+		require.Equal(t, len(idxs), len(pcs))
+		for ii, idx := range idxs {
+			key := cloudletData[idx].Key
+			require.Equal(t, key, pcs[ii].cloudlet.Key, "verify potentialCloudlet[%d] is cloudletData[%d], %s", ii, idx, key.GetKeyString())
+		}
+	}
+
+	// Docker app without serverless config
+	// should find reservD cluster
+	// check potential cloudlets
+	app := getApp()
+	app.Deployment = cloudcommon.DeploymentTypeDocker
+	ai := getAppInst()
+	pclos, err := apis.appInstApi.getPotentialCloudlets(ctx, cctx, ai, app)
+	require.Nil(t, err)
+	verifyCloudlets(pclos, 3, 1, 2, 0)
+	pclus, err := apis.appInstApi.getPotentialClusters(ctx, cctx, ai, app, pclos)
+	require.Nil(t, err)
+	require.Equal(t, 1, len(pclus))
+	require.Equal(t, reservD.Key, pclus[0].existingCluster)
+
+	// Kubernetes app without serverless config
+	// should find reservK cluster
+	app.Deployment = cloudcommon.DeploymentTypeKubernetes
+	pclos, err = apis.appInstApi.getPotentialCloudlets(ctx, cctx, ai, app)
+	require.Nil(t, err)
+	verifyCloudlets(pclos, 3, 1, 2, 0)
+	pclus, err = apis.appInstApi.getPotentialClusters(ctx, cctx, ai, app, pclos)
+	require.Nil(t, err)
+	require.Equal(t, 1, len(pclus))
+	require.Equal(t, reservK.Key, pclus[0].existingCluster)
+
+	// Docker app with serverless config
+	// should find only reservD cluster
```

Comment:
If it finds the same cluster as the test case on line 166, doesn't it need to be checked again?

Reply:
Hm, not sure what you mean. Serverless config is really only for Kubernetes, so results should be the same whether or not docker app specifies allowServerless, that's what this is testing.

Reply:
Ah I see ok - didn't understand whether this is the test case :)

Reply:
let me add a comment to clarify it.